src/toil/provisioners/clusterScaler.py

```diff
@@ -47,193 +46,196 @@
+    given list of jobs. To run the bin-packing algorithm, use the
+    binPack method on a list of job shapes, and use getRequiredNodes
+    to get how many nodes it thinks are required.
+
+    :param list nodeShapes: A list of possible types of nodes that can be launched (as "Shape"s).
+    :param targetTime: The time before which all jobs should at least be started.
     """
-
-    def __init__(self, config, nodeShape, N=1000):
-        # As a prior we start of with 10 jobs each with the default memory, cores, and disk. To
-        # estimate the running time we use the the default wall time of each node allocation,
-        # so that one job will fill the time per node.
-        self.jobShapes = deque(maxlen=N,
-                               iterable=10 * [Shape(wallTime=nodeShape.wallTime,
-                                                    memory=config.defaultMemory,
-                                                    cores=config.defaultCores,
-                                                    disk=config.defaultDisk,
-                                                    preemptable=True)])
-        # Calls to add and getLastNJobShapes may be concurrent
-        self.lock = Lock()
-        # Number of jobs to average over
-        self.N = N
-
-    def add(self, jobShape):
+    def __init__(self, nodeShapes, targetTime=3600):
+        self.nodeShapes = nodeShapes
+        self.targetTime = targetTime
+        self.nodeReservations = {nodeShape:[] for nodeShape in nodeShapes}
+
+    def binPack(self, jobShapes):
+        """Pack a list of jobShapes into the fewest nodes reasonable. Can be run multiple times."""
+        logger.debug('Running bin packing for node shapes %s and %s job(s).', self.nodeShapes, len(jobShapes))
+        # Sort in descending order from largest to smallest. The FFD
+        # like-strategy will pack the jobs in order from longest to
+        # shortest.
+        jobShapes.sort()
+        jobShapes.reverse()
+        assert len(jobShapes) == 0 or jobShapes[0] >= jobShapes[-1]
+        for jS in jobShapes:
+            self.addJobShape(jS)
+
+    def addJobShape(self, jobShape):
         """
-        Adds a job shape as the last completed job.
-        :param Shape jobShape: The memory, core and disk requirements of the completed job
+        Function adds the job to the first node reservation in which it will fit (this
+        is the bin-packing aspect)
         """
-        with self.lock:
-            self.jobShapes.append(jobShape)
+        chosenNodeShape = None
+        for nodeShape in self.nodeShapes:
+            if NodeReservation(nodeShape).fits(jobShape):
+                # This node shape is the first that fits this jobShape
+                chosenNodeShape = nodeShape
+                break
+
+        if chosenNodeShape is None:
+            logger.warning("Couldn't fit job with requirements %r into any nodes in the nodeTypes list." % jobShape)
```

Comment:
I'd vote no, I think, because at least progress could be made via other jobs (or the job may fit onto a previously statically provisioned worker node).